syncer/error.go

```diff
@@ -77,55 +81,159 @@ func originError(err error) error {
-			// only take effect on `ADD INDEX`, no UNIQUE KEY and FOREIGN KEY
-			// UNIQUE KEY may affect correctness, FOREIGN KEY should be filtered.
-			// ref https://github.com/pingcap/tidb/blob/3cdea0dfdf28197ee65545debce8c99e6d2945e3/ddl/ddl_api.go#L1929-L1948.
-			switch v.Specs[0].Constraint.Tp {
-			case ast.ConstraintKey, ast.ConstraintIndex:
-				handle()
-				return nil // ignore the error
+	// for DROP COLUMN with its single-column index, try drop index first then drop column
+	dropColumnF := func(tctx *tcontext.Context, originErr error, ddls []string, index int, conn *DBConn) error {
+		mysqlErr, ok := originError(originErr).(*mysql.MySQLError)
+		if !ok {
+			return originErr
+		}
+		// different version of TiDB has different error message, try to cover most versions
+		if !(mysqlErr.Number == errno.ErrUnsupportedDDLOperation &&
+			strings.Contains(mysqlErr.Message, "drop column") &&
+			strings.Contains(mysqlErr.Message, "with index")) {
+			return originErr
+		}
+		ddl2 := ddls[index]
+		stmt, err2 := parser2.ParseOneStmt(ddl2, "", "")
+		if err2 != nil {
+			return err // return the original error
+		}
+
+		var (
+			schema string
+			table  string
+			col    string
+		)
+		if n, ok := stmt.(*ast.AlterTableStmt); !ok {
+			return originErr
+			// support ALTER TABLE tbl_name DROP
+		} else if len(n.Specs) != 1 {
+			return originErr
+		} else if n.Specs[0].Tp != ast.AlterTableDropColumn {
+			return originErr
+		} else {
+			schema = n.Table.Schema.O
+			table = n.Table.Name.O
+			col = n.Specs[0].OldColumnName.Name.O
+		}
+		tctx.L().Warn("try to fix drop column error", zap.String("DDL", ddl2), log.ShortError(err))
+
+		// check if dependent index is single-column index on this column
+		sql := "SELECT INDEX_NAME FROM information_schema.statistics WHERE TABLE_SCHEMA = ? and TABLE_NAME = ? and COLUMN_NAME = ?"
+		indices, err := conn.querySQL(tctx, sql, schema, table, col)
+		if err != nil {
+			return originErr
+		}
+		var (
+			idx       string
+			idx2Check []string
+			idx2Drop  []string
+			count     int
+		)
+		for indices.Next() {
+			if err := indices.Scan(&idx); err != nil {
+				indices.Close()
+				return originErr
+			}
+			idx2Check = append(idx2Check, idx)
+		}
+		// Close is idempotent, we could close in advance to reuse conn
+		indices.Close()
+
+		sql = "SELECT count(*) FROM information_schema.statistics WHERE TABLE_SCHEMA = ? and TABLE_NAME = ? and INDEX_NAME = ?"
+		for _, idx := range idx2Check {
+			rows, err2 := conn.querySQL(tctx, sql, schema, table, idx)
+			if err2 != nil || !rows.Next() || rows.Scan(&count) != nil || count != 1 {
+				tctx.L().Warn("can't auto drop index", zap.String("index", idx))
+				rows.Close()
+				return originErr
 			}
+			idx2Drop = append(idx2Drop, idx)
+			rows.Close()
+		}
+
+		sqls := make([]string, len(idx2Drop))
+		for i, idx := range idx2Drop {
+			sqls[i] = fmt.Sprintf("ALTER TABLE `%s`.`%s` DROP INDEX `%s`", schema, table, idx)
```

Comment:
how about using `TableName` in github.com/pingcap/tidb-tools/pkg/dbutil

Reply:
changed in https://github.com/pingcap/dm/pull/801/commits/31f6b5a204a67509673f5d1b1d9e19365bae5ad3, and use `dbutil.ColumnName(idx)` to make use of escape backquote of index name (not sure it's OK since function name is `ColumnName`)